src/txn/txn.c

```diff
@@ -1431,33 +1431,21 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
         WT_ERR(__txn_fixup_hs_update(session, hs_cursor));
 
 prepare_verify:
-    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED)) {
-        for (; head_upd != NULL; head_upd = head_upd->next) {
-            /*
-             * Assert if we still have an update from the current transaction that hasn't been
-             * resolved or aborted.
-             */
-            WT_ASSERT_ALWAYS(session,
-              head_upd->txnid == WT_TXN_ABORTED || head_upd->prepare_state == WT_PREPARE_RESOLVED ||
-                head_upd->txnid != txn->id,
-              "Failed to resolve all updates associated with a prepared transaction");
-
-            if (head_upd->txnid == WT_TXN_ABORTED)
-                continue;
-
-            /*
-             * If we restored an update from the history store, it should be the last update on the
-             * chain.
-             */
-            if (!commit && resolve_case == RESOLVE_PREPARE_ON_DISK &&
-              head_upd->type == WT_UPDATE_STANDARD && F_ISSET(head_upd, WT_UPDATE_RESTORED_FROM_HS))
-                WT_ASSERT_ALWAYS(session, head_upd->next == NULL,
-                  "Rolling back a prepared transaction resulted in an invalid update chain");
-
-            /* We have traversed all the non-obsolete updates. */
-            if (WT_UPDATE_DATA_VALUE(head_upd) && __wt_txn_upd_visible_all(session, head_upd))
-                break;
-        }
+    /*
+     * If we are committing a prepared transaction we can check, to some level, that we resolved the
+     * update chain correctly.
+     */
+    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED) && commit) {
```

Comment:
It could be helpful to leave a comment explaining why this check doesn't apply with the rollback case, and also why you don't expect updates to be freed out from underneath in the commit case.

Reply:
Added in https://github.com/wiredtiger/wiredtiger/pull/9350/commits/dd38448ea3c5e8028d224ef1a4204bd56ef49810

Reply:
Nit: added to chain -> added to the chain